### ISSUE TITLE
Auto creation dataset on destination node when is not created

### ISSUE THREAD

**Issue by gusioo** [oetiker/znapzend#206]
Hello!

I found this pull request: [Create datasets on dst when it doesn't exist #194](https://github.com/oetiker/znapzend/pull/194)

I try to test this but I cannot force znapzend to create missing dataset on destination node. I use Znapzend from **master**. Also I use **-AutoCreation** switch.

Does anyone confirm that it works?

Best regards,
Jacek.

**Comment by qx-bug-importer** [comment]
we are just realizing that it does not ...

**Comment by oetiker** [comment]
if you tried before the 4th ...  then maybe it is fixed now ... also, does your user have the appropriate rights ? what error message do you get ?

**Comment by gusioo** [comment]
Hi!

I tested version 0.15.5 but still I can't use --autoCreation switch.

I have two node (**nodeA** and **nodeB**). I create replication task from nodeA to nodeB. On nodeB I have a pool which has not created any dataset.

In znapzend log I have these lines:

`[Thu Apr 14 13:19:40 2016] [warn] destination 'user@192.168.251.192:storage/dataset' does not exist or is offline. will be rechecked every run...`

I try both methods: replication to remote node and replication to another pool on the same node. Do you confirm that it should work in both scenarios?

**Comment by noahmehl** [comment]
I just built from Master, and --autoCreation is not working for me either.  I get:

*** WARNING: destination 'root@1.2.3.4:zones/ae625ad4-072b-11e6-b512-3e1d05defe78' does not exist, will be ignored! ***

**Comment by Wouter0100** [comment]
Hmm, let me check this.

**Comment by Wouter0100** [comment]
The configured dataset in your schedule should exist, only the recurring ones are automatically created.

**Comment by gusioo** [comment]
@Wouter0100 correct me if I am wrong: --autoCreation works only if I have previusly created dataset?

Is it necessary to have datasets created on both nodes before creating task? Or I just need to have a dataset on source node and then dataset on destination will be created automatically.

**Comment by oetiker** [comment]
by default znapzend does automatically create datasets on the dst host in recursive mode. The dataset corresponding to the src dataset must exist though ... 

with the --autoCreation option even the initial dst dataset should get created automatically ...

**Comment by Wouter0100** [comment]
`--autoCreation` only works for datasets which are automatically synced (e.g. the recursive ones). The first dataset (the one where you set the schedule on) has to exist on both destinations on creation.

For example, I have 2 machines, I'd like to backup machine 1 to machine 2. You'll configure the dataset on machine 1 to sync using ZnapZend. This dataset also needs to exist on machine 2. Now, when you've enabled `--autoCreation` on the daemon and `recursive` on the plan, all datasets you'll now create on machine 1 on the configured dataset, will be automatically created on machine 2.

**Comment by oetiker** [comment]
@wouter0100 ... that should work even without the `--autoCreation` option

**Comment by gusioo** [comment]
> with the --autoCreation option even the initial dst dataset should get created automatically ...

So for now it does not work that way. Initial dataset must be created manually.

**Comment by noahmehl** [comment]
@oetiker are you saying it's broken, or working as designed.  I am a bit confused from the thread.

Thanks!

**Comment by oetiker** [comment]
I foolishly accepted a patch providing this feature ... and it was utterly broken ... I thought we had it fixed now, but it seems not to be the case ...

**Comment by Wouter0100** [comment]
It's working as how I designed it, but I see you guys had other expectations from it. Let me fix that :). I'll get back to you with a PR.

**Comment by oetiker** [comment]
@Wouter0100 hihi .. but in recursive mode, new volumes and dataset always were autocreated ...

**Comment by Wouter0100** [comment]
@oetiker hmm, then there should've been something wrong, because that wasn't working on my side :(. I only received an error that the child datasets didn't exist and where skipped. Also in the code I wasn't able to find anything related to `zfs create`.

**Comment by oetiker** [comment]
create is not neccesary because receive will  take care of that automatically  ...

**Comment by oetiker** [comment]
hmmmm ... I guess I should really take a closer look at the code :) things have been working fine for us, but now looking through the code I am not entirely sure why this is the case :)

**Comment by Wouter0100** [comment]
Hmm, that wasn't working, as I remind it correctly, the destination machine returned an error that the dataset didn't exist. Let me find the error and get back to you :).

**Comment by oetiker** [comment]
OK, got it straight again ... what happens is that IF znapzend is in recursive mode, and it finds that the fileset/volume does NOT exist on the destination side, it does not send an incremental snapshot but a full one ... causing the fileset/volume to spring into existance ... so create as such should  never be required.

We just decided when building znapzend that we don't want to have toplevel filesets to show up 'just like that' and thus added code to make sure that only 'sub'-filesets in recursive mode are handled like that, and if a toplevel fileset was missing there would be a complaint.

I'm afraid I was not fully aware of how we implemented it when I reviewed your patch ... so all the autoCreate mode really has todo is todo is to disable znapzends unhappyness about non existing target filesets the rest should be magic

**Comment by Wouter0100** [comment]
Hmm, okay. Let me revert my PR and test it on my server. I'm pretty sure there was something going wrong, that's why I've created that feature.. :P.

**Comment by aaltonenp** [comment]
Will it create the dataset with the same properties as source, or with defaults?

**Comment by Wouter0100** [comment]
@oetiker found it: https://gist.github.com/Wouter0100/78557b6a9f9457008f402e48aab04a4e

**Comment by oetiker** [comment]
if anyone cares to test #211 this would be great

**Comment by gusioo** [comment]
@oetiker I will test it

**Comment by oetiker** [comment]
@Wouter0100 or @gusioo were you successful with the patched version ?

**Comment by gusioo** [comment]
@oetiker I start tests tomorrow

**Comment by gusioo** [comment]
Now works: https://github.com/oetiker/znapzend/commit/c63ec27740a9c10002626cbb00ea8becc3b471d7

**Comment by Harvie** [comment]
It seems to be still broken. I've installed 0.19.0 as described in README.md

`znapzend --noaction --debug --autoCreation --runonce=tank`

I still get follwoing error:

`cannot open 'tank/backup/virt1/vps/vm-101-disk-1': dataset does not exist`

I have following setup:

```
recursive       = on
src             = tank
dst_a           = root@host:tank/backup/virt1
```

And `tank/backup/virt1` exists on dst host.

**Comment by oetiker** [comment]
with --noaction I would expect this to happen

**Comment by Harvie** [comment]
HAHA LOL. Sorry

**Comment by Harvie** [comment]
BTW if i am using recursive replication and i will remove child dataset on SRC, will it get deleted from DST once it gets obsolete according to backup preservation plan?

**Comment by oetiker** [comment]
no datasets will not get removed only old snapshots

**Comment by znek** [comment]
Just recently upgraded to `znapzend-0.22.0` and now it seems that `--autoCreation` ceased to work; here's an excerpt from the log:
```text
…
[2024-06-08 00:03:08.17658] [73736] [warn] sub-destination 'backup:backup/backup/muller/jails/snac' does not exist or is offline; ignoring it for this round... Consider disabling this dataset from znapzend handling.
…
[2024-06-08 00:03:47.73774] [73736] [warn] sub-destination 'backup:backup/backup/muller/roots/14.1-STABLE-06c7e28acca1' does not exist or is offline; ignoring it for this round... Consider disabling this dataset from znapzend handling.
…
[2024-06-08 00:05:55.76495] [73736] [warn] ERROR: suspending cleanup source dataset muller because 2 send task(s) failed:
[2024-06-08 00:05:55.76556] [73736] [warn]  +-->   sub-destination 'backup:backup/backup/muller/jails/snac' does not exist or is offline; ignoring it for this round... Consider disabling this dataset from znapzend handling.
[2024-06-08 00:05:55.76568] [73736] [warn]  +-->   sub-destination 'backup:backup/backup/muller/roots/14.1-STABLE-06c7e28acca1' does not exist or is offline; ignoring it for this round... Consider disabling this dataset from znapzend handling.
[2024-06-08 00:05:55.76586] [73736] [info] done with backupset muller in 354 seconds
```

**Comment by znek** [comment]
This seems to have been already fixed in [d075800]
